### PR TITLE
New role and capabilities manager

### PR DESCRIPTION
**Pull request by incraigulous**
https://www.loom.com/share/e46ed2d5c25041c1897ec3ee9944e531

### CODE REVIEW

**Comment by incraigulous** [comment]
@corsacca See Travis log. any tips on getting the composer install to work for PHP 8?

File: dt-contacts/base-setup.php [hunk 1]
```diff
@@ -127,13 +134,39 @@ public function dt_set_roles_and_permissions( $expected_roles ){
...
         $expected_roles["administrator"]["permissions"] = array_merge( $expected_roles["administrator"]["permissions"], $user_management_permissions );
+        $expected_roles['administrator']['permissions'] = array_merge( $expected_roles["administrator"]["permissions"], $role_caps );
         $expected_roles["administrator"]["permissions"]['manage_dt'] = true;
         $expected_roles["administrator"]["permissions"]['view_project_metrics'] = true;
         $expected_roles["administrator"]["permissions"]["dt_all_admin_contacts"] = true;
 
         return $expected_roles;
     }
 
+    public function dt_set_custom_roles_and_permissions( $roles ) {
```
Comment (corsacca) [inline dt-contacts/base-setup.php:145]: would this be better as part of the function in dt-core/setup-functions.php?

File: dt-core/global-functions.php [hunk 2]
```diff
@@ -905,6 +905,97 @@ function dt_validate_date( string $date, string $format = 'Y-m-d H:i:s' ): bool
...
+        function dd( ...$params ) {
+            foreach ( $params as $param ) {
+                var_dump( $param );
+            }
+
+            exit;
+        }
+    }
+
+    if ( !function_exists( 'dt_sanitize_array' ) ) {
```
Comment (corsacca) [inline dt-core/global-functions.php:921]: This seems like a duplicate of `dt_recursive_sanitize_array`

File: dt-contacts/base-setup.php (hunk 1 above)
Comment (incraigulous) [reply dt-contacts/base-setup.php:145]: @corsacca, I considered that, but that file is only run on admin requests. We need to register custom capabilities on the front end as well. Is there a better place I could put that code that would be available to the front-end as well?

File: dt-core/global-functions.php [hunk 3]
```diff
@@ -905,6 +905,97 @@ function dt_validate_date( string $date, string $format = 'Y-m-d H:i:s' ): bool
         }
     }
 
+    /**
+     * Dump and die
+     */
+    if ( !function_exists( 'dd' ) ) {
```
[the remark this answers is not in the document]
Comment (incraigulous) [reply dt-core/global-functions.php:911]: It stands for dump and die. It dumps a given variable (or variables) and ends execution of the script. It's a convenience utility function you'll find in some PHP frameworks. I'm happy to remove it, but it's so much easier to write: 

```
dd($foo, $bar); 
```

instead of: 

```
var_dump($foo); var_dump($bar); exit;
```

File: dt-core/global-functions.php (hunk 2 above)
Comment (incraigulous) [reply dt-core/global-functions.php]: Removed in favor of `dt_recursive_sanitize_array`

**Comment by incraigulous** [comment]
@corsacca, I've made a round of revisions. Let me know if you would like any further changes.

**Comment by incraigulous** [comment]
@corsacca, any idea why there are tests failing? Those don't seem to be related to my work.

File: dt-contacts/base-setup.php (hunk 1 above)
Comment (corsacca) [reply dt-contacts/base-setup.php:145]: The code in dt-core/setup-functions.php only runs on wp-admin requests so that we aren't running the roles registering on every request.
Once registering is done, the roles settings persist across the all requests.
What part needs to be available on the front end?

File: composer.json [hunk 4]
```diff
@@ -2,6 +2,11 @@
     "require-dev": {
         "dealerdirect/phpcodesniffer-composer-installer": "*",
         "wp-coding-standards/wpcs": "*",
-        "squizlabs/php_codesniffer": "*"
+        "squizlabs/php_codesniffer": "*",
```
Comment (corsacca) [inline composer.json:5]: @incraigulous remove these file changes to fix travis.

File: tests/install-wp-tests.sh [hunk 5]
```diff
@@ -150,6 +150,12 @@ install_db() {
 	mysqladmin create $DB_NAME --user="$DB_USER" --password="$DB_PASS"$EXTRA
 }
 
+install_composer() {
```
Comment (corsacca) [inline tests/install-wp-tests.sh:153]: @incraigulous lets do phpunit tests changes in a different PR

File: tests/bootstrap.php [hunk 6]
```diff
@@ -19,6 +19,10 @@
 // Give access to tests_add_filter() function.
 require_once $_tests_dir . '/includes/functions.php';
 
+if ( file_exists( $_tests_dir . '/vendor/autoload.php' ) ) {
```
Comment (corsacca) [inline tests/bootstrap.php:22]: @incraigulous lets do phpunit tests changes in a different PR

File: dt-core/configuration/class-migration-engine.php [hunk 7]
```diff
@@ -12,7 +12,7 @@
 class Disciple_Tools_Migration_Engine
 {
 
-    public static $migration_number = 46;
+    public static $migration_number = 47;
```
Comment (corsacca) [inline dt-core/configuration/class-migration-engine.php:15]: no longer needed?

**Comment by incraigulous** [comment]
@corsacca, I reverted the phpunit changes. I'll hold on making a pull request for that. I was pretty sure that wasn't the right solution. I was just getting it working for me locally, but I'm sure it's not the right solution (see our slack conversation regarding from around the time of the commit).

File: dt-core/setup-functions.php [hunk 8]
```diff
@@ -70,3 +70,31 @@ function dt_setup_roles_and_permissions(){
...
     }
 }
+
+/**
+ * Add custom roles and permissions from the roles manager.
+ */
+add_filter( 'dt_set_roles_and_permissions', 'dt_setup_custom_roles_and_permissions', 11, 1 );
+
+function dt_setup_custom_roles_and_permissions( $roles ) {
+    $custom_roles = get_option( 'dt_custom_roles' );
```
Comment (corsacca) [inline dt-core/setup-functions.php:80]: @incraigulous i get an invalid arg error if no dt_custom_roles is set.
Suggestion:
`$custom_roles = get_option( 'dt_custom_roles', [] );`

**Comment by corsacca** [comment]
@incraigulous small issue with tab not opening:
tab

**Comment by corsacca** [comment]
@incraigulous 
I'd expect a new role to show up here:
image

**Comment by incraigulous** [comment]
> @incraigulous I'd expect a new role to show up here: image

@corsacca I was able to solve this by loading the roles on the front-end as well as the back-end. You mentioned in a previous comment that you wanted me to set up the custom roles in `dt-core/setup-functions.php`, so I wanted to highlight that I made this change.

**Comment by incraigulous** [comment]
> These are wp caps: image

@corsacca  Odd... those roles aren't showing up for me anymore, though they used to. I assumed you removed them. Can you confirm your WP is updated to the latest version and that they are still showing up for you

Screen Shot 2022-03-17 at 3 54 23 PM
?

**Comment by corsacca** [comment]
@incraigulous you are right about the roles setup needing to be available on each request.
suggestion: https://github.com/thecodezone/disciple-tools-theme/pull/2

For the level 0 to level 10. I checked out the branch on 3 instances, one of them a clean install and see the them show up.

**Comment by incraigulous** [comment]
@corsacca, I just pushed a commit to allow us to remove specific capabilities, namely, level_0 - level_9.

**Comment by corsacca** [comment]
Thank @incraigulous. Would you be able to fix the phpcs check?

**Comment by incraigulous** [comment]
Fixed!

Thank you!

*CRAIG WANN*
Mobile: (918) 641-4282

**Comment by corsacca** [comment]
Thank you @incraigulous !
